[MLIR] add dot offloads with manual tuning support

manupak opened the pull request:
This commit adds dot + pointwise fusion support
along with manual tuning using rocMLIR.

# Usage

## Non-tuning (heuristic mode) default flow

```
MIGRAPHX_ENABLE_MLIR=1 ./bin/migraphx-driver perf --onnx ../models/model_torch_export_fp32.onnx --fill1 input_ids --input-dim @input_ids 1 384
```

This would show warnings that the current compilation is not using tuning db as follows : 

```
Compiling ... 
Reading: ../models/model_torch_export_fp32.onnx
WARNING: MLIR tuning db not found. Please set MIGRAPHX_MLIR_TUNING_DB for optimal performance.
fails to set param ongfx90a:sramecc+:xnack-	-transA false -transB false -g 1 -m 384 -n 768 -k 768 -t f32
WARNING: MLIR tuning db not found. Please set MIGRAPHX_MLIR_TUNING_DB for optimal performance.
....
```

## The manual tuning process

### Obtain the tuning configuration for the model

```
MIGRAPHX_ENABLE_MLIR=1 MIGRAPHX_MLIR_TUNING_CFG=bert.cfg ./bin/migraphx-driver compile --onnx ../models/model_torch_export_fp32.onnx --fill1 input_ids --input-dim @input_ids 1 384
```

This will create at most two files:
1) bert.cfg.gemm
2) bert.cfg.conv (if fused convs were present in the model -- no for bert case)

### Tune using helper python scripts of rocMLIR

1) checkout https://github.com/ROCmSoftwarePlatform/rocMLIR
2) build using the target : ninja check-rocmlir-build-only ci-performance-scripts
3) tune using the configs from the previous step

```
cd build
./bin/tuningRunner.py --op=gemm --configs_file=/var/mkarunar/work/AMDMIGraphX/build/bert.cfg.gemm --output=bt.tsv --verify-mode=none
# if convs were present
./bin/tuningRunner.py --op=conv --configs_file=/var/mkarunar/work/AMDMIGraphX/build/bert.cfg.conv --output=bt.tsv --verify-mode=none
```

Note : even if the .cfg has duplicated lines, the tuning will not perform duplicated tuning and it will append .tsv files. Therefore, one could do this for several models and obtain a single database file.

### Use tuning db with MIGraphX

```
MIGRAPHX_ENABLE_MLIR=1 MIGRAPHX_MLIR_TUNING_DB=/var/mkarunar/work/rocMLIR/build/bt.tsv ./bin/migraphx-driver perf --onnx ../models/model_torch_export_fp32.onnx --fill1 input_ids --input-dim @input_ids 1 384
```

Ideally the above command should not give any warning as we expect all the problem configs to be found in the tuned database.

closes : https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/823 (more like not needed now if we can make this work)

pfultz2 commented on src/targets/gpu/fuse_mlir.cpp:
```diff
@@ -63,43 +64,47 @@ struct mlir_conv
+        return op.compute_shape({inputs[n - 2], inputs[n - 1]});
+    }
+};
+MIGRAPHX_REGISTER_OP(mlir_dot);
```
These could be combined into the same operator:

```cpp
struct mlir_op
{
    operation op;

    template <class Self, class F>
    static auto reflect(Self& self, F f)
    {
        return pack(f(self.op, "op"));
    }

    std::string name() const { return "gpu::mlir_op"; }
    shape compute_shape(std::vector<shape> inputs, const std::vector<module_ref>& mods) const
    {
        check_shapes{inputs, *this}.packed_or_broadcasted();
        if(mods.size() != 1)
            MIGRAPHX_THROW("should have one submodule.");
        if(inputs.size() < 2)
            MIGRAPHX_THROW("should have at least two inputs.");
        auto n = inputs.size();
        return op.compute_shape({inputs[n - 2], inputs[n - 1]});
    }
};
MIGRAPHX_REGISTER_OP(mlir_op);
```

manupak replied:
Yes could be, if thats preferred.

I'd use mlip_binary_op or something because just mlir_op seems too generic for that specific compute_shape implementation.

pfultz2 replied:
> I'd use mlir_binary_op or something because just mlir_op seems too generic for that specific compute_shape implementation.

We can make it work with any number of inputs by checking the number of parameters in the submodule, but we can tackle that later when we do different parameters.

manupak replied:
ok. done.

pfultz2 commented on src/targets/gpu/fuse_mlir.cpp:
```diff
@@ -133,12 +138,57 @@ struct find_conv_pointwise
+        return false;
+    }
+    return true;
+}
```
This matcher is the same as `match::name("dot")`.

pfultz2 commented on src/targets/gpu/fuse_mlir.cpp:
```diff
@@ -133,12 +138,57 @@ struct find_conv_pointwise
+    {
+        auto dot = match::skip(match::name("contiguous"))(is_mlir_dot().bind("gemm_based_op"));
+        return match::name("pointwise")(match::any_of[match::inputs()](dot.bind("x")));
+    }
```
These two matchers can be combined using `match::any_of(match::name("dot"), is_mlir_conv())`:

```cpp
auto matcher() const
{
  auto dot_or_conv = match::skip(match::name("contiguous"))(
      match::any_of(match::name("dot"), is_mlir_conv()).bind("gemm_based_op"));
  return match::name("pointwise")(
      match::any_of[match::inputs()](dot_or_conv.bind("x")));
}
```

manupak replied:
done

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -498,11 +503,24 @@ struct mlir_program
         return ins->get_shape();
     }
 
+    void set_sym_name(const module& m)
```
I think this should just return the symbol name:

```cpp
static std::string get_symbol_name(const module& m)
    {
        for(auto ins : iterator_for(m))
        {
            if(ins->name() == "convolution" or ins->name() == "dot")
            {
                return "mlir_" + ins->name();
            }
        }
        return "main";
    }
```

pfultz2 replied:
And then do `sym_nam = get_symbol_name(m)` in `parse` method.

manupak replied:
done

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -498,11 +503,24 @@ struct mlir_program
+    {
+        for(auto ins : iterator_for(m))
+        {
+            if(ins->name() == "convolution" || ins->name() == "dot")
```
Use `or` instead of `||`.

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -580,12 +599,93 @@ struct mlir_program
+            std::stringstream key(probConfig);
+            std::string prob;
+            int token_count = 0;
+            for(std::string s; getline(key, s, '\t');)
```
Use `split_string(key, '\t')`.

manupak replied:
done

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -580,12 +599,93 @@ struct mlir_program
+                token_count++;
+            }
+
+            if(prob.substr(0, 4) == std::string("conv"))
```
Use `starts_with(prob, "conv")`.

manupak replied:
done

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -580,12 +599,93 @@ struct mlir_program
     problem_params pp;
     std::deque<std::string> strings{};
     std::string target_arch;
+    MlirRockTuningTable tuning_table;
```
This should just be stored in the `get_module_tuned`. I dont see why it needs to be a member variable.

pfultz2 replied:
Actually this should be a static variable so we only read the db file once and not everytime we do a mlir compile.

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -580,12 +599,93 @@ struct mlir_program
+                    std::getline(tuning_db_tsv, perf);
+                    key             = arch + "\t" + prob;
+                    char* key_cstr  = strdup(key.c_str());
+                    char* perf_cstr = strdup(perf.c_str());
```
Dont use `strdup`, this leaks memory.

manupak replied:
oh dear! its not even needed. Apologies Im also picking part of this code.... removed.

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -580,12 +599,93 @@ struct mlir_program
+                while(std::getline(tuning_db_tsv, arch, '\t'))
+                {
+                    std::getline(tuning_db_tsv, prob, '\t');
+                    std::getline(tuning_db_tsv, perf);
```
Use `split_string` here too. Its really hard to follow what is happening here.

pfultz2 commented:
Also, make sure `make tidy` is clean on this PR.

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -580,12 +599,93 @@ struct mlir_program
+                }
+                if(token_count > 1)
+                {
+                    break;
```
Actually, this probably should be `std::find_if` instead of `split_string` since you are only reading the first element, but either way would be fine though.

manupak replied:
done.

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -580,12 +599,93 @@ struct mlir_program
 
+    void tuning_table_create()
+    {
+        tuning_table               = mlirRockTuningTableCreate();
```
Isnt there a `mlirRockTuningTableDestroy` that needs to be called?

manupak replied:
Yes there is .. and it should be called -- now done.

manupak commented on src/targets/gpu/mlir.cpp:
```diff
@@ -580,12 +599,93 @@ struct mlir_program
+
+            if(prob.substr(0, 4) == std::string("conv"))
+            {
+                tuning_cfg_path = tuning_cfg_path + ".conv";
```
done.

manupak commented:
Thanks @pfultz2 for detailed review. much appreciated.. (Im also picking part of this code up).

I've addressed all comments except the tuning table lifetime concern. From what I can see, mlir compiler is a static set of functions to be called through out compiling the whole graph.

> Actually this should be a static variable so we only read the db file once and not everytime we do a mlir compile.

Whilst I agree with you we should only be creating the tuning table once per compilation, I dont see an immediate way to to such an init and destroy -- also personally not a fan of global static here but then I dont see non-invasive way (i.e. outside the build ifdef guard) to inject mlir things to compile flow.  

Any suggestions ?

(I've added the destroy tuning table functionality for now -- so it doesn't live beyond a single mlir_program)

pfultz2 commented on src/targets/gpu/mlir.cpp:
```diff
@@ -641,6 +697,34 @@ void adjust_param_shapes(module& m, const std::vector<instruction_ref>& inputs)
     }
 }
 
+static void update_tuning_table(MlirRockTuningTable& tuning_table)
```
This function should return the `tuning_table` and be named `create_tuning_table`.

```cpp
static mlir_tuning_table create_tuning_table()
{
...
}
```

pfultz2 commented:
> Whilst I agree with you we should only be creating the tuning table once per compilation, I dont see an immediate way to do such an init and destroy -- also personally not a fan of a static var here but then I dont see non-invasive way (i.e. outside the build ifdef guard) to inject mlir things to compile flow. If you insist, I could add that.

Just add it as a static variable to `get_module_tuned`:

```cpp
bool get_module_tuned() const
{
    static auto tuning_table = create_tuning_table();
    if(!mlirRockTuningSetFromTable(tuning_table.get(), mmodule.get()))
    {
        const char* prob_config = mlirRockTuningGetKey(tuning_table, mmodule.get());
        std::stringstream key(prob_config);
        std::cerr << "fails to set param on" << prob_config << std::endl;
        dump_tuning_cfg(prob_config);
        return false;
    }
    return true;
}
```

manupak commented on src/targets/gpu/mlir.cpp:
```diff
@@ -654,10 +738,13 @@ code_object_op compile_mlir(const context&, module m, const std::vector<instruct
         std::cout << mlir_print(&mlirOperationPrint, mod_op) << std::endl;
-    auto co   = mp.compile();
+    auto co = mp.compile(tuning_table);
+    mlirRockTuningTableDestroy(tuning_table);
```
done. (now I know about MIGRAPHX_MANAGE_MLIR_HANDLE)

manupak commented on src/targets/gpu/mlir.cpp:
```diff
@@ -654,10 +738,13 @@ code_object_op compile_mlir(const context&, module m, const std::vector<instruct
     mlir_program mp;
     mp.find_target();
     mp.parse(m);
+    MlirRockTuningTable tuning_table = mlirRockTuningTableCreate();
```
done

manupak commented:
> Just add it as a static variable to `get_module_tuned`:

Done.

This change requires a bump to rocMLIR commit as well (which is done here). 
Will the CI build the docker images ?

causten commented:
> 

Lately we've been clearing the cache manually.  Not sure why it hasn't been doing that automatically

sjw36 commented on src/targets/gpu/fuse_mlir.cpp:
```diff
@@ -79,27 +80,27 @@ MIGRAPHX_PRED_MATCHER(is_mlir_conv, instruction_ref ins)
-               return not contains({"@literal", "@param", "@return", "convolution", "add", "relu"},
-                                   i.name());
+               return not contains(
+                   {"@literal", "@param", "@return", "convolution", "dot", "add", "relu"},
```
Does this restrict our fusion flow to only add/relu elementwise ops?

manupak replied:
Historically it has been the case -- the actual reasons might predate me :). 
Thus, its not a change in this PR (see the red bits), we can look to expand the whitelist in a future PR -- but I dont think we can remove it altogether (because all elementwise ops of migraphx does not have a lowering in rocMLIR)

sjw36 reviewed (approved):
Looks good overall. Just 1 question.

manupak commented:
@causten I did clear the caches w.r.t this branch (though it wasn't that much roughly 300ish mbs).
Although, if you look at GH actions they just fail in post-run docker caching step -- IIUC, the checks have passed.

What are the next steps here ?
